data/src/main/java/com/linkedin/data/schema/annotation/DataSchemaRichContextTraverser.java

```diff
@@ -505,6 +508,16 @@ TraverserContext getNextContext(String nextTraversePathComponent, String nextSch
       return nextContext;
     }
 
+    public DataSchema getOriginalTopLevelSchema()
+    {
+      return _originalTopLevelSchema;
+    }
+
+    void setOriginalTopLevelSchema(DataSchema originalTopLevelSchema)
```

Comment:
You want customized visitor to modify context?

Reply:
(1) some of the methods still have access control, they can't set all the field of the context.
(2) but yes, the visitor implementation could have access "currentSchema" and "topLevelSchema" now. I weighted between "hide implementation for transparency" vs "expose implementation for optimization", and feel that visitor implementation should have access to the details.

Reply:
> You want customized visitor to modify context?

(1) some of the methods still have access control, they can't set all the field of the context.
(2) but yes, the visitor implementation could have access "currentSchema" and "topLevelSchema" now. I weighted between "hide implementation for transparency" vs "expose implementation for optimization", and feel that visitor implementation should have access to the details.

Reply:
Make all the setters private and then create a new method for getFirstContext? That way you don't have to expose the setters outside of this class.

Infact, you should consider defining an interface for the context and then making this implementation class private.

Reply:
As @karthikbalasub commented, I may prefer that we make all setters private.